### PR TITLE
Improve search performance at song select when a collection is selected

### PR DESCRIPTION
**Pull request by peppy**
Without this, there's a large overhead to do a realm-live `Contains` search when a collection is selected.

This may also help considerably alleviate https://github.com/ppy/osu/discussions/27298#discussioncomment-8552508 as we will be performing the native realm search much less.

I've tested that this handled items being added/removed from collections correctly, so it seems like an oversight from the outset:

2024-02-23 11 19 29@2x

Closes https://github.com/ppy/osu/issues/27322.

### CODE REVIEW

**Comment by bdach** [comment]
How was this tested? Is there a collection I can use quickly, or am I supposed to manufacture my own?

File: osu.Game/Screens/Select/FilterControl.cs [hunk 1]
```diff
@@ -64,7 +65,7 @@ public FilterCriteria CreateCriteria()
                 Sort = sortMode.Value,
                 AllowConvertedBeatmaps = showConverted.Value,
                 Ruleset = ruleset.Value,
-                CollectionBeatmapMD5Hashes = collectionDropdown.Current.Value?.Collection?.PerformRead(c => c.BeatmapMD5Hashes)
+                CollectionBeatmapMD5Hashes = collectionDropdown.Current.Value?.Collection?.PerformRead(c => c.BeatmapMD5Hashes).ToList()
```
Comment (bdach) [inline osu.Game/Screens/Select/FilterControl.cs:68]: At the point this is being forcibly materialised, might as well do it to a type that is designed for this sort of usage? By which I mean

```suggestion
                CollectionBeatmapMD5Hashes = collectionDropdown.Current.Value?.Collection?.PerformRead(c => c.BeatmapMD5Hashes).ToHashSet()
```

or even

```suggestion
                CollectionBeatmapMD5Hashes = collectionDropdown.Current.Value?.Collection?.PerformRead(c => c.BeatmapMD5Hashes).ToImmutableHashSet()
```

I'm not sure if things are going to be smart enough to work with just that given that `CollectionBeatmapMD5Hashes` is just an `IEnumerable<string>`, but maybe? (and the type could always be changed if using hashset has no gains when calling `.Contains()` through `IEnumerable`)

Comment (bdach) [reply osu.Game/Screens/Select/FilterControl.cs:68]: In fact, after testing with the database provided [here](https://github.com/ppy/osu/issues/27322#issuecomment-1960943814), as far as I can tell, using `.ToList()` actually makes the performance of this _much worse_ in extreme cases unless `ImmutableHashSet` is used!

| master | this PR | this PR but `ImmutableHashSet` |
| :-: | :-: | :-: |
| master | list | hashset |

Note that the list variant takes _literal minutes_ to complete.

Comment (peppy) [reply osu.Game/Screens/Select/FilterControl.cs:68]: Ah yep, that's an oversight. Was waiting to have data to test with to confirm results.

**Review by bdach — APPROVED** [review]

**Review by frenzibyte — APPROVED** [review]

**Pull request merged**